lib/srv/db/redis/engine.go

```diff
@@ -195,7 +196,18 @@ func (e *Engine) getNewClientFn(ctx context.Context, sessionCtx *common.Session)
 	}
 
 	return func(username, password string) (redis.UniversalClient, error) {
-		redisClient, err := newClient(ctx, connectionOptions, tlsConfig, username, password)
+		var onConnect onClientConnectFunc
+		switch {
+		// If password is provided by client.
+		case password != "":
+			onConnect = authWithPasswordOnConnect(username, password)
+
+		// If database user is one of managed users.
+		case apiutils.SliceContainsStr(sessionCtx.Database.GetManagedUsers(), sessionCtx.DatabaseUser):
+			onConnect = fetchUserPasswordOnConnect(sessionCtx, e.Users)
```

Comment:
Changing the user password does not terminate the existing session. This is mainly for Cluster mode in case password changes when connecting to a shard at a later time.

As suggested, I will refactor this part in a separate function and document that.